testing/upload_validation.ts

```diff
@@ -933,137 +943,141 @@ const genericObjectSchema: z.ZodTypeAny = z.lazy(() =>
               imagePropertySchema.codaType as ValueHintType,
-            )
-          ) {
-            context.addIssue({
-              code: z.ZodIssueCode.custom,
-              path: ['imageProperty'],
-              message: `The "imageProperty" field name "${schema.imageProperty}" must refer to a "ValueType.String" property with a "ValueHintType.ImageAttachment" or "ValueHintType.ImageReference" "codaType".`,
-            });
-            return;
-          }
-        }
+            ),
+          `must refer to a "ValueType.String" property with a "ValueHintType.ImageAttachment" or "ValueHintType.ImageReference" "codaType".`,
+        );
       };
-      const validateDescriptionProperty = () => {
-        if (schema.descriptionProperty) {
-          if (!(schema.descriptionProperty in schema.properties)) {
-            context.addIssue({
-              code: z.ZodIssueCode.custom,
-              path: ['descriptionProperty'],
-              message: `The "descriptionProperty" field name "${schema.descriptionProperty}" does not exist in the "properties" object.`,
-            });
-            return;
-          }
-
-          const descriptionPropertySchema = schema.properties[schema.descriptionProperty];
-          if (
-            descriptionPropertySchema.type !== ValueType.String &&
-            (descriptionPropertySchema.type !== ValueType.Array ||
-              descriptionPropertySchema.items.type !== ValueType.String)
-          ) {
-            context.addIssue({
-              code: z.ZodIssueCode.custom,
-              path: ['descriptionProperty'],
-              message: `The "descriptionProperty" field name "${schema.descriptionProperty}" must refer to a "ValueType.String" property or array of strings.`,
-            });
-            return;
-          }
-        }
+      const validateSnippetProperty = () => {
+        return validateProperty(
+          'snippetProperty',
+          snippetPropertySchema =>
+            snippetPropertySchema.type === ValueType.String ||
+            (snippetPropertySchema.type === ValueType.Array && snippetPropertySchema.items.type === ValueType.String),
+          `must refer to a "ValueType.String" property or array of strings.`,
+        );
       };
       const validateLinkProperty = () => {
-        if (schema.linkProperty) {
-          if (!(schema.linkProperty in schema.properties)) {
+        return validateProperty(
+          'linkProperty',
+          linkPropertySchema =>
+            linkPropertySchema.type === ValueType.String && linkPropertySchema.codaType === ValueHintType.Url,
+          `must refer to a "ValueType.String" property with a "ValueHintType.Url" "codaType".`,
+        );
+      };
+
+      validateTitleProperty();
+      validateLinkProperty();
+      validateImageProperty();
+      validateSnippetProperty();
+      (schema.subtitleProperties || []).forEach((f, i) => {
+        if (typeof f === 'string') {
```

Comment:
Is this missing a validation case for when f is a PropertyIdentifierDetails?

Reply:
oops good catch i forgot to refactor this one. done